compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,178 @@ static bool consumesInputOp(NamedSequenceOp op) {
+      }
+    }
+  }
+
+  // Process actions.
+  for (auto action : foreachMatch.getActions()) {
+    if (auto actionSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(action)) {
+      if (auto actionOp = SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+              module, actionSymRef)) {
+        namedSequenceToForeachMatch[actionOp] = foreachMatch;
+      }
+    }
+  }
+}
+
+// Collects tuning specs (`NamedSequenceOp`s) from all inner modules within the
+// given input model, and organizes them into a `TuningSpecsToMerge` struct.
+// - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+//     - Are not named `__kernel_config`.
+//     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+// - `namedSequenceToForeachMatch`: Maps `NamedSequenceOp`s used as matchers or
+//    actions to their corresponding `ForeachMatchOp`.
+static TuningSpecsToMerge collectTuningSpecsToMerge(ModuleOp module) {
+  TuningSpecsToMerge tuningSpecs;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() != kKernelConfigSpecName ||
+          !namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+        tuningSpecs.namedSequenceOpsToMove.push_back(namedSequenceOp);
+        continue;
+      }
+
+      // Extract the single ForeachMatchOp inside namedSequenceOp.
+      ForeachMatchOp foreachMatch =
+          *namedSequenceOp.getOps<ForeachMatchOp>().begin();
+      assert(foreachMatch &&
+             "ForeachMatch should exist in the `__kernel_config`.");
+
+      updateForeachMatchMappings(foreachMatch, innerModule,
+                                 tuningSpecs.namedSequenceToForeachMatch);
+    }
+  }
+
+  return tuningSpecs;
+}
+
+// Renames a `NamedSequenceOp` to resolve name conflicts caused by merging
+// tuning specs, moves it to the top-level module, and updates its occurrences
+// in corresponding `ForeachMatchOp` operations.
+static void
+updateNamedSequenceOp(NamedSequenceOp op, ModuleOp module, OpBuilder &builder,
+                      llvm::StringMap<unsigned> &specNameCounts,
+                      llvm::DenseMap<NamedSequenceOp, ForeachMatchOp>
+                          &namedSequenceToForeachMatch) {
+  StringRef specName = op.getSymName();
+  std::string newSpecName = getUniqueSpecName(specName, specNameCounts);
+  op.setSymName(newSpecName);
+
+  // Skip updating ForeachMatchOp if the NamedSequenceOp is not used in it
+  // or its name has not changed.
+  if (!namedSequenceToForeachMatch.contains(op) || newSpecName == specName) {
+    op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+    return;
+  }
+
+  ForeachMatchOp foreachMatchOp = namedSequenceToForeachMatch[op];
+
+  // Helper function for updating matchers or actions in associated
+  // `foreachMatchOp` instances.
+  auto updateSymbol = [&](Attribute attr) -> SymbolRefAttr {
```

Comment:
Update would suggest this performs some actions and modifies the attribute. But because attributes are immutable, I'd expect it to create a new one instead. Maybe `getUpdatedSymbol` instead?